### PR TITLE
fix: FireBaseAnalytics.setCurrentScreen() on Fragment may not be called

### LINKED ISSUE #500: [assigned] FireBaseAnalytics.setCurrentScreen() on Fragment may not be called
## Overview (Required)
(日本語で失礼します)

一部のFragmentで`FireBaseAnalytics.setCurrentScreen`が呼ばれないケースがあります。
そのFragmentとは、`AllSessionsFragment`や`SearchSessionsFragment`と、ViewPagerの中の最初のFragmentです。
一例として`SearchSessionsFragment`のライフサイクルは以下のようになります。

0. アプリを起動
0. bottom navigationで検索を選択
0. `SearchSessionsFragment`インスタンス生成
0. `setUserVisibleHint()`が呼ばれるが **ログは送れない**
0. `onAttach()`が呼ばれる
0. `onActivityCreated()`が呼ばれる（fireBaseAnalyticsインスタンスの初期化）
0. トピックタブ(`SearchTopicsFragment`)に移動、トピックタブのログは送れる
0. 再度セッションタブ(`SearchSessionsFragment`)に移動、この時の**ログは送れる**

現状`setUserVisibleHint()`でログを送っていますが、 上記の 3 の時点だと`fireBaseAnalytics` はまだnullなので実行されません。
たとえ`fireBaseAnalytics`の初期化タイミングを早めて実行できたとしても、まだこのときはAttachされておらず`getActivity()`がnullなので`setCurrentScreen()`を実行できません。

代わりに`ViewPager.OnPageChangeListener` を使って試そうとしましたが、あまりいいやり方にできなかったのでissueにしました 🤔 

## Links
- https://developer.android.com/reference/android/app/Fragment.html#setUserVisibleHint(boolean)

### PR DESCRIPTION
**Pull request by satoshun**
## Issue
- close #500 

## Overview (Required)

setCurrentScreen isn't called from Fragment of ViewPager. cuz fireBaseAnalytics instance doesn't create at setUserVisibleHint.(only first fragment into ViewPager)
So, extends setPrimaryItem and send fireBaseAnalytics event in setPrimaryItem method. setPrimaryItem method, Fragment guarantee to be active.

and it's pr including small dirty hack(FragmentStateNullablePagerAdapter.java). 
base PagerAdapter class has `PagerAdapter#setPrimaryItem(@NonNull ViewGroup, int, @NonNull Object)`  method, but actually `PagerAdapter#setPrimaryItem(@NonNull ViewGroup, int, @Nullable Object)`. so i create FragmentStateNullablePagerAdapter class and override setPrimaryItem method with `@Nullable`.

---

(日本語でも)

setCurrentScreenがViewPager内のFramgentから呼ばれていませんでした。setUserVisibleHintメソッドが呼ばれたタイミングでfireBaseAnalyticsが生成されていないことがあったためです。なので、setPrimaryItemメソッド内でfireBaseAnalyticsイベントを呼び出すことにしました。

## Screenshot
Before | After
:--: | :--:
 |

### CODE REVIEW

File: app/src/main/java/io/github/droidkaigi/confsched2018/presentation/FragmentStateNullablePagerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import android.support.annotation.Nullable;
+import android.support.v4.app.FragmentManager;
+import android.support.v4.app.FragmentStatePagerAdapter;
+import android.view.ViewGroup;
+
+/**
+ * memo:
+ * FragmentStateNullablePagerAdapter needs to be written in Kotlin.
+ * cuz, setPrimaryItem signature is #setPrimaryItem(@NonNull ViewGroup, int, @NonNull Object),
+ * but Object come in null, so kotlin occurs NPE crash
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/presentation/FragmentStateNullablePagerAdapter.java]: OMG 😰

File: app/src/main/java/io/github/droidkaigi/confsched2018/presentation/sessions/SessionsFragment.kt [hunk 2]
```diff
@@ -101,8 +103,12 @@ class SessionsFragment : Fragment(), Injectable, Findable, OnReselectedListener
 }
 
 class SessionsViewPagerAdapter(
-        fragmentManager: FragmentManager
-) : FragmentStatePagerAdapter(fragmentManager) {
+        fragmentManager: FragmentManager,
+        private val activity: Activity
+) : FragmentStateNullablePagerAdapter(fragmentManager) {
+
+    private var currentFragment: Fragment? = null
```
Comment (takahirom) [inline app/src/main/java/io/github/droidkaigi/confsched2018/presentation/sessions/SessionsFragment.kt]: How about using `Delegates.observable`?

```
var currentFragment: Fragment? by Delegates.observable(null) { _, old, new ->
  if(new != old) {
      ...
  }
}
```

**Comment by takahirom** [comment]
I reviewed!
Almost LGTM! 👍

File: app/src/main/java/io/github/droidkaigi/confsched2018/presentation/FragmentStateNullablePagerAdapter.java (hunk 1 above)
Comment (satoshun) [reply app/src/main/java/io/github/droidkaigi/confsched2018/presentation/FragmentStateNullablePagerAdapter.java]: i have no idea more elegant way(java has `NonNull` annotation but actually Nullable pattern) 😢😢😢

Comment (takahirom) [reply app/src/main/java/io/github/droidkaigi/confsched2018/presentation/FragmentStateNullablePagerAdapter.java]: It's ok 🙆

File: app/src/main/java/io/github/droidkaigi/confsched2018/presentation/sessions/SessionsFragment.kt (hunk 2 above)
Comment (satoshun) [reply app/src/main/java/io/github/droidkaigi/confsched2018/presentation/sessions/SessionsFragment.kt]: completely work! thx comment 🙏

**Comment by satoshun** [comment]
fixed 🙏 thx useful comment!

**Comment by takahirom** [comment]
LGTM 👍

**Comment by takahirom** [comment]
Thank you for fixing difficult problem 💯

**Pull request merged**